src/hashfunctions.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Datastructures: GAP package providing common datastructures.
+ * Licensed under the GPL 2 or later.
+ *
+ * hashfun: various hash functions
+ */
+
+#ifndef HASHFUNCTIONS_H
+#define HASHFUNCTIONS_H
+
+#include "datastructures.h"
+#include "src/intfuncs.h"
+
+extern struct DatastructuresModule HashFunctionsModule;
+
+
+// Hash two integers together
+static inline UInt HashCombine2(UInt hash1, UInt hash2);
+
+// Transform a UInt into a GAP intermediate integer, shrinking
+// the size of the number as required
+static inline Obj HashValueToObjInt(UInt hash);
+
+// We do not use INTOBJ_INT here, because our UInt might not
+// fit inside an INTOBJ, but we do not care about throwing the
+// top bits away
+static inline Obj HashValueToObjInt(UInt hash)
+{
+    return (Obj)((hash << 2) + 0x01);
```

Comment:
This may produce invalid INTOBJ objects, though, as the two top bits *must* be equal, to allow for overflow checks.